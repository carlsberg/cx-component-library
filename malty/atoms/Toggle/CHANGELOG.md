# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.1.0 (2024-05-31)

### Bug Fixes

- **DSM-314:** label clickable & flickering test ([#254](https://github.com/CarlsbergGBS/cx-component-library/issues/254)) ([b2fbf0a](https://github.com/CarlsbergGBS/cx-component-library/commit/b2fbf0a698ff7833fbffeefa3046e457058165c8))
- **DSM-314:** toggle flickering ([#273](https://github.com/CarlsbergGBS/cx-component-library/issues/273)) ([4b1ffd8](https://github.com/CarlsbergGBS/cx-component-library/commit/4b1ffd8e353c4ce26479ff8314dd10ce09563e08))
- Remove typography provider from every component ([#383](https://github.com/CarlsbergGBS/cx-component-library/issues/383)) ([5d05680](https://github.com/CarlsbergGBS/cx-component-library/commit/5d05680e40d3da5bb3042bd8191427a70606afc3))
- reverted changes ([#396](https://github.com/CarlsbergGBS/cx-component-library/issues/396)) ([ecfcf57](https://github.com/CarlsbergGBS/cx-component-library/commit/ecfcf573863253fea16e24764b6f452c3cc4f215))
- Run auto fixable lint based on new `eslint` configuration ([#501](https://github.com/CarlsbergGBS/cx-component-library/issues/501)) ([9ee94cc](https://github.com/CarlsbergGBS/cx-component-library/commit/9ee94cc2cd407f717e62d8857b5f20a74e7bbec4))
- **toggle:** [DSM-703] prevent small jump when toggling off ([#721](https://github.com/CarlsbergGBS/cx-component-library/issues/721)) ([8b2b1c6](https://github.com/CarlsbergGBS/cx-component-library/commit/8b2b1c6bfa1d9073a6bc496a2cb1aa03bcc0d9ba))
- **toggle:** [DSM-728] update some styles according to design specifications ([#723](https://github.com/CarlsbergGBS/cx-component-library/issues/723)) ([4541d78](https://github.com/CarlsbergGBS/cx-component-library/commit/4541d78c814c31b46461223b63bdb8c36781d4e5))

### Features

- add tsconfig files and update package.json files for each component ([3bfc644](https://github.com/CarlsbergGBS/cx-component-library/commit/3bfc644e1cfc9dbb7cf7a0469e25fce055b53240))
- additional stories ([#291](https://github.com/CarlsbergGBS/cx-component-library/issues/291)) ([438af3b](https://github.com/CarlsbergGBS/cx-component-library/commit/438af3b969e0f5b4b908436ee58b93c4b9ddc97c))
- change organization name ([a0aa428](https://github.com/CarlsbergGBS/cx-component-library/commit/a0aa428f43138af5707a3ff4c0e36b7c056f02e6))
- **DSM-228:** Added rating component ([#379](https://github.com/CarlsbergGBS/cx-component-library/issues/379)) ([ab47354](https://github.com/CarlsbergGBS/cx-component-library/commit/ab47354db803fb051a1f87d66ee34966698df8bd))
- **DSM-248:** Headline HTML tags ([#184](https://github.com/CarlsbergGBS/cx-component-library/issues/184)) ([305f5e0](https://github.com/CarlsbergGBS/cx-component-library/commit/305f5e0fa92d5813f4877b338ad827a8c347d452))
- **DSM-368:** Required prop & label component ([#384](https://github.com/CarlsbergGBS/cx-component-library/issues/384)) ([983bd78](https://github.com/CarlsbergGBS/cx-component-library/commit/983bd78c8f526a8faa2df601bf918b40219820f2))
- **DSM-439:** Change font-family behavior ([#470](https://github.com/CarlsbergGBS/cx-component-library/issues/470)) ([ab6dad4](https://github.com/CarlsbergGBS/cx-component-library/commit/ab6dad414c42c4343de3de6abaa4cd3ff98cfca1))
- **global:** [DSM-784] add space props in all components ([#842](https://github.com/CarlsbergGBS/cx-component-library/issues/842)) ([9f03f59](https://github.com/CarlsbergGBS/cx-component-library/commit/9f03f59e7518a99abcea06bacbee989d1ddb0465))
- **global:** migrate from bit to lerna ([a0110c8](https://github.com/CarlsbergGBS/cx-component-library/commit/a0110c8831370dc762c193b17cc593eed381f990))
- **global:** Remove `react/jsx-props-no-spreading` comments ([#558](https://github.com/CarlsbergGBS/cx-component-library/issues/558)) ([823444c](https://github.com/CarlsbergGBS/cx-component-library/commit/823444ce0800bc7f49b584de59d247eb8497f559))
- **global:** update storybook files to be compliant with v7 ([#970](https://github.com/CarlsbergGBS/cx-component-library/issues/970)) ([c61d4b4](https://github.com/CarlsbergGBS/cx-component-library/commit/c61d4b40a0755becf942ad3f28758a159f8c54e4))
- integrate visual testing ([#925](https://github.com/CarlsbergGBS/cx-component-library/issues/925)) ([35dfc1c](https://github.com/CarlsbergGBS/cx-component-library/commit/35dfc1cce6b76d5d2661d552f9fb25477d3a57f7))
- remove empty composition files ([f8173f4](https://github.com/CarlsbergGBS/cx-component-library/commit/f8173f4a2ecbf80bb7b6ffe848c023ae31819c2d))
- **toggle:** [DSM-798] add tests ([#775](https://github.com/CarlsbergGBS/cx-component-library/issues/775)) ([3c47017](https://github.com/CarlsbergGBS/cx-component-library/commit/3c47017d97607916720663c1cc392a885377389e))
- Use new `eslint-config-carlsberg` with streamlined way to create components ([#559](https://github.com/CarlsbergGBS/cx-component-library/issues/559)) ([a459332](https://github.com/CarlsbergGBS/cx-component-library/commit/a45933215bc0b523220743f17d988d7f1ad5fce6))
