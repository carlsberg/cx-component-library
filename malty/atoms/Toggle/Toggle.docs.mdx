---
labels: ['react', 'typescript', 'toggle']
description: 'A toggle component.'
---
