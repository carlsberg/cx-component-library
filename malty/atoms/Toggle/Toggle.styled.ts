import { Label } from '@carlsberggroup/malty.atoms.label';
import styled, { css } from 'styled-components';

export const StyledError = styled.label`
  font-family: inherit;
  color: ${({ theme }) => theme.colors.colours.system.fail.value};
  font-size: ${({ theme }) => theme.typography.desktop.text.tiny_default['font-size'].value};
  font-weight: bold;
  line-height: ${({ theme }) => theme.typography.desktop.text.tiny_default['line-height'].value};
  letter-spacing: 0;
`;

export const StyledLabelWrapper = styled.div<{
  disabled?: boolean;
}>`
  align-items: center;
  display: inline-block;
  display: flex;
  // using hardcoded values due to not having token value and design team thinking on implementation with icons
  height: 14px;
  position: relative;
  width: auto;
  ${({ disabled }) =>
    disabled &&
    css`
      pointer-events: none;
    `}
  &:focus,
  &:focus-visible {
    box-shadow: none;
    outline: none;
  }
`;
export const StyledToggleSwitch = styled.div`
  position: relative;
  display: inline-block;
  width: ${({ theme }) => theme.sizes.m.value};
  // using hardcoded values due to not having token value and design team thinking on implementation with icons
  height: 14px;
  &:focus,
  &:focus-visible {
    box-shadow: none;
    outline: none;
  }
`;
export const StyledLabel = styled(Label)<{
  disabled?: boolean;
  required?: boolean;
}>`
  font-weight: normal;
  font-size: ${({ theme }) => theme.typography.desktop.text['medium-small_default']['font-size'].value};
  line-height: ${({ theme }) => theme.typography.desktop.text['medium-small_default']['line-height'].value};
  margin-left: ${({ theme }) => theme.sizes['2xs'].value};
  margin-bottom: 0;
`;
export const StyledInput = styled.input<{
  disabled?: boolean;
}>`
  display: none;
  &:checked + .switch::before {
    transform: translateX(14px);
    background-color: ${({ theme }) => theme.colors.colours.default.white.value};
    border: none;
    top: 1px;
    transition: transform 0.3s ease;
  }
  &:checked + .switch {
    background-color: ${({ theme }) => theme.colors.colours.default['digital-black'].value};

    &:hover {
      ${({ theme }) =>
        theme &&
        css`
          background-color: ${theme.colors.colours.overlay['digital-black'][75].value};
          border: ${theme.borders['border-2px--solid']['border-width'].value}
            ${theme.borders['border-2px--solid']['border-style'].value} transparent;
        `}
    }
  }
  &:focus,
  &:focus-visible {
    box-shadow: none;
    outline: none;
  }
  ${({ disabled, theme }) =>
    disabled &&
    css`
      &:checked + .switch {
        background-color: ${theme.colors.colours.system['disable-light-theme'].value};
        pointer-events: none;
        border: ${theme.borders['border-2px--solid']['border-width'].value}
          ${theme.borders['border-2px--solid']['border-style'].value} transparent;
      }
    `}
`;

export const StyledSwitch = styled.span<{
  disabled?: boolean;
}>`
  position: absolute;
  cursor: pointer;
  background-color: ${({ theme }) => theme.colors.colours.default.white.value};
  border-radius: ${({ theme }) => theme.sizes.m.value};
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  transition: all 0.3s ease;

  ${({ theme }) =>
    theme &&
    css`
      border: ${theme.borders['border-2px--solid']['border-width'].value}
        ${theme.borders['border-2px--solid']['border-style'].value}
        ${theme.colors.colours.default['digital-black'].value};
      &:hover {
        border: ${theme.borders['border-2px--solid']['border-width'].value}
          ${theme.borders['border-2px--solid']['border-style'].value}
          ${theme.colors.colours.overlay['digital-black'][75].value};
        &:before {
          border: ${theme.borders['border-2px--solid']['border-width'].value}
            ${theme.borders['border-2px--solid']['border-style'].value}
            ${theme.colors.colours.overlay['digital-black'][75].value};
        }
      }
    `}

  &:before {
    position: absolute;
    content: '';
    left: -2px;
    top: -1px;
    width: ${({ theme }) => theme.sizes['2xs'].value};
    height: ${({ theme }) => theme.sizes['2xs'].value};
    background-color: ${({ theme }) => theme.colors.colours.default.white.value};

    border-radius: 50%;
    transition: all 0.3s ease;
    ${({ theme }) =>
      theme &&
      css`
        border: ${theme.borders['border-2px--solid']['border-width'].value}
          ${theme.borders['border-2px--solid']['border-style'].value}
          ${theme.colors.colours.default['digital-black'].value};
      `}
  }
  ${({ disabled, theme }) =>
    disabled &&
    css`
      &:before {
        border: ${theme.borders['border-2px--solid']['border-width'].value}
          ${theme.borders['border-2px--solid']['border-style'].value}
          ${theme.colors.colours.system['disable-light-theme'].value};
        pointer-events: none;
      }
      pointer-events: none;
      border: ${theme.borders['border-2px--solid']['border-width'].value}
        ${theme.borders['border-2px--solid']['border-style'].value}
        ${theme.colors.colours.system['disable-light-theme'].value};
    `}
`;
