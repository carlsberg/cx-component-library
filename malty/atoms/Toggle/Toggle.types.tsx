export interface ToggleProps extends React.HTMLAttributes<HTMLInputElement> {
  onValueChange: (value: boolean) => void;
  label?: string;
  disabled?: boolean;
  checked: boolean;
  error?: string;
  required?: boolean;
}
