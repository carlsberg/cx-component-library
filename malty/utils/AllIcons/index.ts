export { allIcons, allIconsStoryOptions } from './AllIconsMap';
