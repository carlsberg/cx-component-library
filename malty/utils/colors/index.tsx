const rgbToHex = (value: string): string =>
  value.replace(/rgb\((\d+),\s*(\d+),\s*(\d+)\)/gi, (_, r, g, b) => {
    const toHex = (n: string) => parseInt(n, 10).toString(16).padStart(2, '0');
    return `#${toHex(r)}${toHex(g)}${toHex(b)}`;
  });

export { rgbToHex };
