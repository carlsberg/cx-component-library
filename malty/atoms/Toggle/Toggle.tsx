import { globalTheme as defaultTheme, TypographyProvider } from '@carlsberggroup/malty.theme.malty-theme-provider';
import React, { useEffect, useMemo, useState } from 'react';
import { v4 as uuid } from 'uuid';
import {
  StyledError,
  StyledInput,
  StyledLabel,
  StyledLabelWrapper,
  StyledSwitch,
  StyledToggleSwitch
} from './Toggle.styled';
import { ToggleProps } from './Toggle.types';

export const Toggle = ({ disabled, onValueChange, label, checked = false, error, ...props }: ToggleProps) => {
  const theme = defaultTheme;
  const [stateChecked, setStateChecked] = useState(checked);
  const id = useMemo(() => uuid(), []);

  const handleToggle = () => {
    onValueChange(!stateChecked);
  };
  useEffect(() => {
    setStateChecked(checked);
  }, [checked]);

  return (
    <TypographyProvider>
      <StyledLabelWrapper disabled={disabled} onClick={handleToggle} theme={theme}>
        <StyledToggleSwitch id={id} theme={theme}>
          <StyledInput
            id={id}
            theme={theme}
            disabled={disabled}
            type="checkbox"
            checked={stateChecked}
            onChange={handleToggle}
            // eslint-disable-next-line react/jsx-props-no-spreading
            {...props}
          />
          <StyledSwitch id={id} theme={theme} disabled={disabled} className="switch" />
        </StyledToggleSwitch>

        <StyledLabel htmlFor={id} theme={theme} disabled={disabled}>
          {label}
        </StyledLabel>
      </StyledLabelWrapper>
      {error && <StyledError theme={theme}>{error}</StyledError>}
    </TypographyProvider>
  );
};
