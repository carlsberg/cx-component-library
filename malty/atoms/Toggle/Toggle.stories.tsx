import { generateStorybookSpacing } from '@carlsberggbs/malty.utils.space';
import { Meta, StoryObj } from '@storybook/react';
import React, { useState } from 'react';
import { Toggle } from './Toggle';
import { ToggleProps } from './Toggle.types';

const ToggleComponent = ({ checked, ...props }: ToggleProps) => {
  const [stateChecked, setStateChecked] = useState(checked);

  return <Toggle {...props} checked={stateChecked} onValueChange={() => setStateChecked(!stateChecked)} />;
};

const meta: Meta<ToggleProps> = {
  title: 'Forms/Toggle',
  component: Toggle,
  parameters: {
    importObject: 'Toggle',
    importPath: '@carlsberggbs/malty.atoms.toggle'
  },
  render: (args) => <ToggleComponent {...args} />,
  argTypes: {
    label: {
      control: 'text',
      description: 'Progress indicating text'
    },
    checked: {
      control: 'none',
      description: 'If toggle is on (checked) or off (unchecked)',
      table: {
        category: 'State'
      }
    },
    disabled: {
      control: 'boolean',
      description: 'Toggle state, when disabled it is read-only',
      table: {
        category: 'State'
      }
    },
    error: {
      description: 'Error message to be displayed when error is present',
      control: 'text'
    },
    onValueChange: {
      description: 'Function to be executed when toggle state changes',
      table: {
        category: 'Events'
      }
    },
    required: {
      control: 'boolean',
      description: 'Makes the toogle required to fill'
    },
    dataTestId: {
      description: 'Toggle data-testid',
      control: 'text'
    },
    ...generateStorybookSpacing()
  }
};

type Story = StoryObj<ToggleProps>;

export const Base: Story = {
  args: {
    label: 'Toggle label',
    disabled: false,
    required: false,
    dataTestId: 'toggle'
  }
};

export const Disabled: Story = {
  args: {
    ...Base.args,
    disabled: true
  }
};

export const Checked: Story = {
  args: {
    ...Base.args,
    checked: true
  }
};

export default meta;
