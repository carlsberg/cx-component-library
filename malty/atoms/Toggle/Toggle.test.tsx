import { render } from '@carlsberggbs/malty.utils.test';
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { Toggle } from './Toggle';

jest.mock('uuid', () => ({ v4: () => '00000000-0000-0000-0000-000000000000' }));

const mockFn = jest.fn();

describe('toggle', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render elements', () => {
    render(<Toggle label="Label text" error="Error text" dataTestId="toggle" checked onValueChange={mockFn} />);
    expect(screen.getByText('Label text')).toBeInTheDocument();
    expect(screen.getByText('Error text')).toBeInTheDocument();
    expect(screen.getByTestId('toggle-input')).toBeInTheDocument();
  });

  it('should call function on click', () => {
    render(<Toggle label="Label text" error="Error text" checked onValueChange={mockFn} />);
    const toggle = screen.getByText('Label text');

    userEvent.click(toggle);

    expect(mockFn).toHaveBeenCalledTimes(1);
  });

  it('should be disabled', () => {
    render(
      <Toggle label="Label text" error="Error text" dataTestId="toggle" checked disabled onValueChange={mockFn} />
    );

    // TODO: transform this into "getByText(labelText)" on the following ticket https://carlsberggbs.atlassian.net/browse/DSM-822
    const toggle = screen.getByTestId('toggle-input');

    expect(toggle).toBeDisabled();
  });

  it('should be checked', () => {
    render(<Toggle label="Label text" error="Error text" dataTestId="toggle" checked onValueChange={mockFn} />);
    const toggle = screen.getByTestId('toggle-input');

    expect(toggle).toBeChecked();
  });

  it('should be unchecked', () => {
    render(<Toggle label="Label text" error="Error text" dataTestId="toggle" disabled onValueChange={mockFn} />);
    const toggle = screen.getByTestId('toggle-input');

    expect(toggle).not.toBeChecked();
  });
});
