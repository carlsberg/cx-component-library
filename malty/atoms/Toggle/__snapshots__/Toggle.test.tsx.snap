// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toggle matches snapshot 1`] = `
<span
  className="sc-bczRLJ hROJNc"
>
  <label
    className="sc-dkzDqf gWKurO"
  >
    <div
      className="sc-hKMtZM GTTLb"
    >
      <input
        checked={false}
        className="sc-jSMfEi jwiPhP"
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="sc-gKXOVf kHFmuB switch"
      />
    </div>
    <label
      className="sc-eCYdqJ dXEYVD"
    >
      Label text
    </label>
  </label>
</span>
`;
